### PR TITLE
Fix: Incorrectly warning about ignored files

### LINKED ISSUE #3649: Incorrectly warning about ignored files
After globs were added to the cli input in 5a599c70a98eaed0d43a8f28690921ef4ad0702d, my command to lint rule docs has broken:

```
DEBUG=eslint* node_modules/.bin/eslint --plugin markdown --no-eslintrc --config conf/rule-doc-config.json docs/rules/quote-props.md
  eslint:cli Running on files +0ms
  eslint:config Using command line config conf/rule-doc-config.json +3ms
  eslint:cli-engine Linting complete in: 3ms +10ms

/home/ian/Local_code/eslint/eslint/docs/rules/quote-props.md
  0:0  warning  File ignored because of your .eslintignore file. Use --no-ignore to override

✖ 1 problem (0 errors, 1 warning)
```

Compare to my output if I revert to the commit prior to globs being added:

```
▶ git reset --hard 37bb3159f748463861ba9aadf704e76af3c2537d
HEAD is now at 37bb315 Merge pull request #3518 from ev0rtex/update-expression-rule-logic

▶ DEBUG=eslint* node_modules/.bin/eslint --plugin markdown --no-eslintrc --config conf/rule-doc-config.json docs/rules/quote-props.md
  eslint:cli Running on files +0ms
  eslint:config Using command line config conf/rule-doc-config.json +3ms
  eslint:cli-engine Processing docs/rules/quote-props.md +10ms
  eslint:cli-engine Linting docs/rules/quote-props.md +1ms
  eslint:config Constructing config for /home/ian/Local_code/eslint/eslint/docs/rules/quote-props.md +0ms
  eslint:config Not using .eslintrc or package.json files +0ms
  eslint:config Merging command line config file +1ms
  eslint:config Merging command line environment settings +0ms
  eslint:config Merging command line plugins +0ms
  eslint:cli-engine Load plugin markdown +4ms

docs/rules/quote-props.md
    7:20  error  Unexpected token ;
  130:6   error  Unexpected string

✖ 2 problems (2 errors, 0 warnings)
```

Something appears to have changed in how files are determined to be ignored.  Furthermore, if I follow the warning's advice and add `--no-ignore`, the warning is not shown, but the file is still not linted:

```
▶ DEBUG=eslint* node_modules/.bin/eslint --plugin markdown --no-ignore --no-eslintrc --config conf/rule-doc-config.json docs/rules/quote-props.md
  eslint:cli Running on files +0ms
  eslint:config Using command line config conf/rule-doc-config.json +2ms
  eslint:cli-engine Linting complete in: 2ms +9ms
```

### PR DESCRIPTION
**Pull request by BYK**
Fixes #3649

### CODE REVIEW

**Comment by nzakas** [comment]
Can you rebase?

**Comment by BYK** [comment]
@nzakas done

File: lib/cli-engine.js [hunk 1]
```diff
@@ -496,15 +517,19 @@ CLIEngine.prototype = {
             results.push(res);
         }
 
-        files.forEach(function(pattern) {
-            glob.sync(pattern, globOptions).forEach(executeOnFile);
+        patterns.forEach(function(pattern) {
+            if (shell.test("-f", pattern) && !ignoredPaths.contains(pattern)) {
```
Comment (gyandeeps) [inline lib/cli-engine.js]: Why not just use `fs` insead of `shell`. And that way we can get rid of the shell dependency and move it back to devDependencies.

Comment (nzakas) [inline lib/cli-engine.js]: We use ShellJS in a few places already.

Comment (gyandeeps) [inline lib/cli-engine.js]: shelljs is only used in here at 2 spots where we can easily use `fs`. By doing that we can move `shelljs` back to `devDependencies` sections

Comment (nzakas) [inline lib/cli-engine.js]: This was already discussed in a previous issue. `fs.exists` is deprecated and is being removed., so we can't rely on it. What is your alternative?

Comment (gyandeeps) [inline lib/cli-engine.js]: we can use `fs.statSync(pattern).isFile()` to check for if its a file.
https://nodejs.org/api/fs.html#fs_class_fs_stats

Comment (BYK) [inline lib/cli-engine.js]: That requires a try-catch block. This is the reason why @nzakas suggested using shelljs.

Comment (gyandeeps) [inline lib/cli-engine.js]: got it. Thanks
@byk What is the use case where `ignoredPaths.contains(pattern)` will be `true` under the condition that its a file?

Comment (BYK) [inline lib/cli-engine.js]: Not sure what you are asking. Can you clarify? 

This new if block I added checks if the pattern passed is actually a file and not ignored by the `.eslintignore` file. If that is the case it skips glob and directly executes the file.

Comment (gyandeeps) [inline lib/cli-engine.js]: What do i need to have in my eslintignore file to make these 2 conditions true
- `shell.test("-f", pattern)` === true
- `ignoredPaths.contains(pattern)` === true

Comment (BYK) [inline lib/cli-engine.js]: This case: https://github.com/eslint/eslint/pull/3666/files#diff-6e3261f1362e4e13d9d684e4cae1a127R398

**Pull request merged**